src/operator/nn/cudnn/cudnn_batch_norm-inl.h

```diff
@@ -228,7 +228,7 @@ class CuDNNBatchNormOp {
         &a,
         &b,
         &a,
-        req[cudnnbatchnorm::kGamma] == kWriteTo ? &b: &b_add,
+        req[cudnnbatchnorm::kGamma] == kAddTo ? &b_add : &b,
```

Comment:
Here, we just test with `req[cudnnbatchnorm::kGamma] == kAddTo`. We won't be able to set the gradient accumulation appropriately if `req[cudnnbatchnorm::kGamma] == kWriteTo && req[cudnnbatchnorm::kBeta] == kAddTo`.

Reply:
@zhreshold I believe there are some `grad_req = add` issues in BN. Would you know if there are cases that `gamma.grad_req = write` and `beta.grad_req = add`? If not, we may just raise an error in the OP if it happens.

Reply:
I don't think there's any use case I can imagine, raising an exception is acceptable

Reply:
@sxjscience @zhreshold Thanks for your review! I will add the grad_req check and an unittest for gradient accumulation.

Reply:
`gamma.grad_req = kAddTo` is not supported in [Naive CPU Implementation](https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/batch_norm.cc#L304), [Naive GPU one](https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/batch_norm.cu#L383) and [MKLDNN one](https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h#L419).

Reply:
I believe it shows a larger problem of the BN layer...

Reply:
The gradient of gamma on CPU is wrong when `grad_req` is True.
```python
import mxnet as mx
from mxnet.gluon import nn

N = 1
C = 3
H = W = 2
block = nn.BatchNorm() 
block.collect_params().initialize()
block.collect_params().setattr('grad_req', 'add')

for i in range(1):
    with mx.autograd.record():
        x = mx.nd.arange(N*C*H*W).reshape((N, C, H, W))
        y = block(x)
        loss = (y * y).sum() 
    loss.backward()
print(block.gamma.grad())
```

Reply:
I guess this is somehow out-of-the-scope of this PR. How about to create an issue and we can have another PR to fix the problem?

Reply:
@sxjscience I will create an issue.